### PR TITLE
Register Newspack Patterns like Core instead of custom solution

### PR DESCRIPTION
**Pull request by thomasguillot**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-theme/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR updates the way we offer our Patterns to users. Previously we used a custom solution. Now we can rely on Core's functions to offer them.

_Note: The previews for the Homepage Posts are quite slow (at least on local). Not quite sure what we can do about it._

### How to test the changes in this Pull Request:

1. Switch to this branch
2. Test all the different patterns. Do you see them, do they work?

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by jeffersonrabb** [comment]
@thomasguillot did you notice if there's a way to register patterns only for certain custom post types? In the original implementation we only showed Homepage Posts patterns on Pages which made the pattern selection much cleaner in Campaigns. This is especially true given the slowness loading previews of Homepage Posts patterns.

**Comment by thomasguillot** [comment]
> @thomasguillot did you notice if there's a way to register patterns only for certain custom post types? In the original implementation we only showed Homepage Posts patterns on Pages which made the pattern selection much cleaner in Campaigns. This is especially true given the slowness loading previews of Homepage Posts patterns.

Good point.
Done in https://github.com/Automattic/newspack-blocks/pull/639/commits/e415b04d639a9068ca0d89a5d5476a21205a86c4

**Review by jeffersonrabb — CHANGES_REQUESTED** [review]
The post type filtering isn't working quite right. When a new post/page/CPT is first created, none of the patterns are available. If the post is saved and the page refreshed, then it works (because the post ID can be extracted from the URL). Also, the mailing list patterns aren't showing up under any circumstances. I'd guess this is because `has_block()` is being called before the Mailchimp block has been registered.

**Comment by jeffersonrabb** [comment]
@dkoo I'm interested to get your take on this work, as it's similar to work you've done with Newspack Listings patterns.

**Comment by dkoo** [comment]
> @dkoo I'm interested to get your take on this work, as it's similar to work you've done with Newspack Listings patterns.

So the issue at question is to register block patterns only for certain post types, right? I have a solution in Listings which is kind of messy, but works for now. The problem is that block patterns must be registered pretty early on (`admin_init`), which is before functions like `get_post_type` or even the global post object are available. So in Listings, I tried getting the post type from the raw request params like so: https://github.com/Automattic/newspack-listings/pull/23/files#diff-e7aef70bd283581544beeaf8ae8ff938e3437abdf2793a22c8ac2d5b3858bacaR154-R168

Not sure if this is the best method, but it was the only way I was able to make it work.

**Comment by dkoo** [comment]
@thomasguillot @jeffersonrabb I see now that this PR attempts to do the same thing I did in Listings to determine post type. I think we'll need to figure out a different method to get that info before the global `$post` object is available. `has_block` also depends on the global `$post` object, so that's probably why that function also isn't working as expected.

One thing we could try is to register block patterns for all posts to begin with, then use [`unregister_block_pattern`](https://developer.wordpress.org/block-editor/developers/block-api/block-patterns/#unregister_block_pattern_category) for the posts and post types we don't want on a later hook. That might let us run `has_block` on the content hook, too.

**Comment by thomasguillot** [comment]
@dkoo @jeffersonrabb I have removed the post type check and the Mailchimp check. I'm OK having the various patterns available to all the post types. I think it's up to the user common sense.

**Comment by dkoo** [comment]
@thomasguillot Sorry about the delay, I was working on 372a30d to do the selective block pattern registration by post type, but your comment beat me to it. I was having trouble with new posts until implementing 21d1821. I had to figure this out because the same issue exists in the Listings plugin. Anyway, take a look and see if it works for you. What it does:

* Moves class PHP files to an `includes` folder to match the other Newspack plugins; directory structure
* Turns the patterns file into a class for consistency
* This is the key: uses `admin_init` instead of `init` to register patterns—this hook runs a bit later than `init`, so the `$_REQUEST` params are available by then.
* Handles new posts: the `$_REQUEST['post']` param doesn't exist for those, but `$_REQUEST['post_type']` does, so we can just use that instead of trying to get the post type by a non-existent post ID.

I tested with various post types and it seems to work for me for both new posts and editing existing posts. How I tested:

1. Create a new post: confirm that the "Newspack Donations" and "Newspack Subscribe" categories and patterns exist.
2. Edit an existing post: confirm the same.
3. Create a new page: confirm that the "Newspack Homepage Posts", Newspack Donations" and "Newspack Subscribe" categories and patterns exist.
4. Edit an existing page: confirm the same.
5. Create a new prompt: confirm that the "Newspack Homepage Posts", Newspack Donations" and "Newspack Subscribe" categories and patterns exist.
6. Edit an existing prompt: confirm the same.
7. Create a new post of another CPT (such as a listing type), and confirm that none of these patterns are available.
8. Edit an existing CPT and confirm the same.

**Comment by dkoo** [comment]
@jeffersonrabb or @adekbadek, could use a review from one of you since I contributed code to this PR.

**Comment by thomasguillot** [comment]
It's great, I like it. Thank you @dkoo

**Review by adekbadek — APPROVED** [review]

File: includes/class-newspack-blocks-patterns.php [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+					$donations_configured = true;
+				}
+			}
+
+			$block_patterns = array();
+
+			// Donations.
+			if (
+				$donations_configured &&
+				( null === $post_type || in_array( $post_type, [ 'page', 'post', 'newspack_popups_cpt' ], true ) )
```
[the remark this answers is not in the document]
Comment (dkoo) [reply includes/class-newspack-blocks-patterns.php]: Good suggestion—added in 3e61bb7.

**Review by adekbadek — APPROVED** [review]

**Pull request merged**